src/workflow/cache_manager.rs

```diff
@@ -0,0 +1,154 @@
+use crate::workflow::WorkflowCachingPolicy;
+use lru::LruCache;
+use std::collections::HashSet;
+
+pub(crate) struct WorkflowCacheManager {
+    /// LRU cache of run IDs.
+    cache: LruCache<String, bool>,
+    /// Defines eviction behavior from the cache when tasks are completed.
+    eviction_policy: WorkflowCachingPolicy,
+    /// List of run ids that should be evicted upon completion.
+    pending_evictions: HashSet<String>,
+}
+
+impl WorkflowCacheManager {
+    pub fn new(policy: WorkflowCachingPolicy) -> Self {
+        let cap = match policy {
+            WorkflowCachingPolicy::Sticky {
+                max_cached_workflows,
+            } => max_cached_workflows,
+            _ => 0,
+        };
+        Self {
+            cache: LruCache::new(cap),
+            eviction_policy: policy,
+            pending_evictions: Default::default(),
+        }
+    }
+
+    /// Touches record associated with run id in the cache.
+    /// Once cache reaches capacity, overflow records will be added into the pending evictions map
+    /// in the LRU order. This function can be safely called on the same run id multiple times.
+    pub fn touch(&mut self, run_id: &str) {
+        if self.cache.len() < self.cache.cap() {
+            // Blindly add a record into the cache, since it still has capacity.
+            self.cache.put(run_id.to_owned(), true);
+            // Remove pending eviction if one exists since there's been recent activity for the run id.
+            self.pending_evictions.remove(run_id);
+        } else if self.cache.cap() != 0 {
+            // If cache is full we want to append least recently used item to pending_evictions.
+            if let Some(lru) = self.cache.peek_lru() {
+                self.pending_evictions.insert(lru.0.to_owned());
+            }
+            // Run id should be removed from pending evictions since it has just been touched.
+            self.pending_evictions.remove(run_id);
+            // Adding a record into the cache.
+            self.cache.put(run_id.to_owned(), true);
+        } else {
+            // Cache size is set to 0, we should add run_id to the list of pending evictions.
+            self.pending_evictions.insert(run_id.to_owned());
+        }
+    }
+
+    /// Returns all pending evictions currently recorded.
+    pub fn get_pending_evictions(&self) -> HashSet<String> {
+        self.pending_evictions.clone()
+    }
```

Comment:
It would be a good idea to avoid cloning here. This should return by reference. Probably returning a set is not a relevant detail either. It could return `&impl Iterator<Item = String>` (or `&str`, not sure which) and then the method contents are just `self.pending_evictions.iter()`

Reply:
This code is no longer needed in the recent revision.